### PR TITLE
[11.0][IMP/FIX] hr_timesheet_sheet (partial)

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Split of https://github.com/OCA/hr-timesheet/pull/188.

### CODE REVIEW

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I'm not sure I follow why there's an iteration over `self` here

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:451]: won't task.id resolve to False if task is empty?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:472]: won't project.id resolve to false if project is empty?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: let's extract this constant somewhere

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:451]: yes. What's the problem?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: ?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:451]: I mean, won't `'task_id': task.id,` be enough?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: `'/'` is used in a lot of places, just an idea to make it a constant

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:451]: if `task` is not an empty record but is a `False` or `None`, then it breaks. It's common to do this `and` to avoid possible breaks just in case.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:472]: same answer as above

**Comment by MiquelRForgeFlow** [comment]
Comments attended.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:413]: [edited]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Do we really need to pass these as params? What's be benefit of that?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:413]: This is called on a single timesheet only, this can be expressed as
```
@api.multi
def button_add_line(self):
    self.ensure_one()

    if self.state == 'draft':
       self.add_line()
```

w/o need to pass details

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Optional inheritance. Just in case you want to make another module that depends of this one, and wants other functionalities, then it's better to have options like this, to pass other projects or tasks.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: It seems to me that w/o parameters it's much easier to extend just add_line, otherwise I'd have to override both `button_add_line` and `add_line` if I'd like to add extra `add_line_smth`

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:413]: I don't like to put "ensure ones" in "api multi" methods. Just my taste and preference.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: I differ with you. I think with parameters it enables more possible customization. If not, `_prepare_empty_analytic_line` method is restricted by what happens in `button_add_line` method, but  `add_line` method cannot influence.

**Comment by MiquelRForgeFlow** [comment]
@astirpe

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:413]: I got that, yet I don't see how exactly this method may be called for multiple objects?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Got it, makes sense, can we do something like _preapare_new_line_params?

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: I don't understand. What method are you mentioning?

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]:
```

    @api.multi
    def button_add_line(self):
        for rec in self:
            if rec.state == 'draft':
                rec.add_line(rec.project_id)
                rec.reset_add_line()
        return True

def add_line(self, project_id, ):
        if self.add_line_project_id:
            values = self._prepare_empty_analytic_line(project_id)
            name_line = self._get_new_line_name(values)
    
```

Something like this

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Why: https://github.com/OCA/hr-timesheet/pull/154/files#diff-c470854eb5351f22cc8515bd7cfe525eR67
https://github.com/OCA/hr-timesheet/pull/154/files#diff-c470854eb5351f22cc8515bd7cfe525eR80

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Still I don't understand. If in `add_line` method I do `name_line = self._get_new_line_name()` instead of `name_line = self._get_line_name(project, task)`, that would go against the intention of giving more possible customization.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: `self._get_new_line_name(values)`?

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Anyhow, it's not related to this improvement and should be discussed separately, I guess.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Let me try to re-express my vision: I'm trying to limit number of places that pass project + task to functions, in order to be able to extend project + task + something w/o need to inject into multiple places.

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: `values` is a dictionary. I know what you are trying, but your experiments until now are not good. So, I think the best way to proceed is to try to merge this PR as it is now and later if you want you can make another with your thing, because if not, it would seem that this thing you are trying is really blocking this PR.

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:416]: Not blocking at all, I'll just review the commits to be portable to v12, and approve :) Totally agree on route

**Comment by astirpe** [comment]
Is this PR containing just a refactoring? Or does it change/fix something functionally?

**Comment by MiquelRForgeFlow** [comment]
The commits with `[FIX]` are fixes and the commits with `[IMP]` are not functional changes.

**Comment by astirpe** [comment]
What is the functional use case fixed by commit "[FIX] Correctly handle partial amounts of timesheets"?
How can I check/verify it?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Can you rename parameter `timesheet` to `timesheets`, to make it consistent with the change you made in line 448: `self.clean_timesheets(timesheets)` ?

Last part of the method would be:

```
        if len(repeated) > 1 and self.id:
            return repeated.merge_timesheets()
        return timesheets
```


Can't we just avoid the merging of lines in `def action_timesheet_draft()`? See https://github.com/OCA/hr-timesheet/issues/197

(comment edited)

**Review by alexey-pelykh — APPROVED** [review]
Code LGTM 👍 , yet `49db0c6128d4bb7a709a670a2b77286bc10b2f6b` won't be able to make it to v12

**Comment by MiquelRForgeFlow** [comment]
@astirpe  Buff, I am too busy to find the use case this following days. When I found it time ago, I was playing with merging several timesheets with same name, ones with positive unit_amount and others with negative unit_amount. Thus, I rechecked all the possible use cases one by one and fixed the wrong ones.

**Comment by MiquelRForgeFlow** [comment]
> won't be able to make it to v12

@alexey-pelykh won't be because it can't or because you don't want?  😏😀

**Comment by alexey-pelykh** [comment]
@mreficent I'm not that kind of bad person :) Won't make it as-is due to current code differs already, base is different

**Comment by MiquelRForgeFlow** [comment]
@alexey-pelykh ok :+1:

**Review by astirpe — CHANGES_REQUESTED** [review]

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Renamed. Also your issue has been fixed in a new commit. The problem was the api multi of the `action_timesheet_draft` method.

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: No, sorry, the problem is in the `_get_timesheet_sheet_lines_domain` method.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I think that https://github.com/OCA/hr-timesheet/commit/640d9289d5f9ece39cb93e2bc144b8cb892ffdde#diff-a48ce9b933da452cb1910c0e256932c7 makes the `clean_timesheets()` not necessary anymore here.


`clean_timesheets()` does two things:

1. sets the proper `sheet_id` in the lines in case it's wrong or not set;
2. merges the lines with description = `/`


Point 1 should be non needed anymore after https://github.com/OCA/hr-timesheet/commit/640d9289d5f9ece39cb93e2bc144b8cb892ffdde#diff-a48ce9b933da452cb1910c0e256932c7 because the computation of `sheet_id` is now in the create and the write of the lines. If I'm not wrong, this part of code is now obsolete and can be safely removed.

Point 2 could be also removed IMO when resetting to draft, as proposed in #197

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Yeah, removed the `clean_timesheets` in the "set to draft" method

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Thanks!

**Review by astirpe — CHANGES_REQUESTED** [review]
Reproduced in Runbot:


 - Create some lines in the details tab. Save.
 - Click on Edit.
 - In the details tab, remove the lines one by one (click on the trash bin icon).

When attempting to remove the last line, all the previously deleted lines reappear again.

**Comment by astirpe** [comment]
Sorry, the issue I mentioned also occurs in current version of the module.
So it's not a regression caused by this PR. I'm going to open a dedicated issue for that.

**Review by astirpe — APPROVED** [review]
Fixes https://github.com/OCA/hr-timesheet/issues/197

**Comment by MiquelRForgeFlow** [comment]
> So it's not a regression caused by this PR. I'm going to open a dedicated issue for that.

No problem. I think it's better to fix/improve this module PR by PR, step by step. If we do all that in one sole PR, it will never get merged 🤣

**Pull request merged**
